### ISSUE TITLE
Pytorch DataParallel did not work on p3.8xlarge

### ISSUE THREAD

**Issue by AtsunoriFujita** [aws/sagemaker-python-sdk#1509]
**Describe the bug**
I ran a learning job using ml.p3.xlarge. 
However, the execution time was almost the same as p3.2xlarge.

applied this code.
`if torch.cuda.device_count() > 1:
     print("Let's use", torch.cuda.device_count(), "GPUs!")
     # dim = 0 [30, xxx] -> [10, ...], [10, ...], [10, ...] on 3 GPUs
     model = nn.DataParallel(model)`

and I got 'Let's use 4 GPUs!'

**To reproduce**
A clear, step-by-step set of instructions to reproduce the bug.

execution time
p3.8xlarge: 4094
p3.2xlarge: 3534

**Expected behavior**
It should be 3-4 times faster than p3.2xlarge.

**Screenshots or logs**
If applicable, add screenshots or logs to help explain your problem.

**System information**
A description of your system. Please provide:
- **SageMaker Python SDK version**:
- **Framework name (eg. PyTorch) or algorithm (eg. KMeans)**:Pytorch
- **Framework version**:1.4.0
- **Python version**:
- **CPU or GPU**:GPU
- **Custom Docker image (Y/N)**:N

**Additional context**

**Comment by Roshrini** [comment]
Hi, Can you provide steps to reproduce the issue? e.g. training script/dataset etc.

**Comment by AtsunoriFujita** [comment]
`
device = torch.device("cuda" if torch.cuda.is_available() else "cpu") 

if __name__ == '__main__':

    parser = argparse.ArgumentParser()

    # Data and model checkpoints directories
    parser.add_argument('--batch_size', type=int, default=16, metavar='N',
                        help='input batch size for training (default: 16)')
    parser.add_argument('--test_batch-size', type=int, default=64, metavar='N',
                        help='input batch size for testing (default: 64)')
    parser.add_argument('--epochs', type=int, default=10, metavar='N',
                        help='number of epochs to train (default: 10)')
    parser.add_argument('--lr', type=float, default=0.01, metavar='LR',
                        help='learning rate (default: 0.01)')
    parser.add_argument('--backend', type=str, default=None,
                        help='backend for distributed training (tcp, gloo on cpu and gloo, nccl on gpu)')

    # Container environment
    parser.add_argument('--hosts', type=list, default=json.loads(os.environ['SM_HOSTS']))
    parser.add_argument('--current-host', type=str, default=os.environ['SM_CURRENT_HOST'])
    parser.add_argument('--model_dir', type=str, default=os.environ['SM_MODEL_DIR'])
    parser.add_argument('--training', type=str, default=os.environ['SM_CHANNEL_TRAINING'])
    parser.add_argument('--num_gpus', type=int, default=os.environ['SM_NUM_GPUS'])

    args = parser.parse_args()

    #is_distributed = torch.cuda.device_count() > 1
    seed_torch(seed=seed, is_distributed=is_distributed)
    is_distributed = torch.cuda.device_count() > 1

    train_df = pd.read_csv('./train_df.csv')
    val_df = pd.read_csv('./val_df.csv')

    img_size = 512
    AUGMENTATIONS_TRAIN = Compose([
        VerticalFlip(p=0.5),
        HorizontalFlip(p=0.5),
        ToFloat(max_value=255),
        ToTensor()
    ], p=1)

    AUGMENTATIONS_TEST = Compose([
        ToFloat(max_value=255),
        ToTensor()
    ], p=1)

    batch_size = args.batch_size
    num_workers = 8

    train_dataset = Alaska2Dataset(train_df,
                                   augmentations=AUGMENTATIONS_TRAIN)
    valid_dataset = Alaska2Dataset(val_df,
                                   augmentations=AUGMENTATIONS_TEST)

    train_loader = torch.utils.data.DataLoader(train_dataset,
                                               batch_size=batch_size,
                                               num_workers=num_workers,
                                               shuffle=True)

    valid_loader = torch.utils.data.DataLoader(valid_dataset,
                                               batch_size=batch_size,
                                               num_workers=num_workers,
                                               shuffle=False)

    model = Net(num_classes=len(class_labels)).to(device)
    model = model.to(device)

    if torch.cuda.device_count() > 1:
        print("Let's use", torch.cuda.device_count(), "GPUs!")
        # dim = 0 [30, xxx] -> [10, ...], [10, ...], [10, ...] on 3 GPUs
        model = nn.DataParallel(model)

    train(model, train_loader, valid_loader, is_distributed, args.model_dir, args.epochs, batch_size)    
`

**Comment by AtsunoriFujita** [comment]
I used Lustre(file size is 30GB and The number of files is about 300,000).

`sess = sagemaker.Session()

estimator = PyTorch(entry_point="train.py",
                    source_dir='source',
                    role=role,
                    framework_version='1.4.0',
                    train_volume_size=100,
                    subnets=['subnet1', 'subnet2'],
                    security_group_ids=['sg-1', 'sg-2'],
                    train_instance_count=2,
                    train_instance_type='ml.p3.8xlarge',
                    sagemaker_session=sess,
                    hyperparameters={
                        'batch_size':16,
                        'lr': 0.0001,
                        'epochs': 1,
                        'backend': 'nccl'
                    }
                   )`

**Comment by Roshrini** [comment]
Were you able to track gpu usage using nvidia-smi to see if all the gpus are getting used? Also this seems to be general PT usage error, you can post dataparallel query in PT forum to get more eyes. https://discuss.pytorch.org/

**Comment by choidongyeon** [comment]
Hi @AtsunoriFujita, do you have any updates on this?

**Comment by AtsunoriFujita** [comment]
Sorry for late reply.
The batch size did not seem to be large enough to do parallel processing.
It was faster on my local environment (RTX2080Ti * 2), but on V100 * 4 I needed to increase the batch size even more.
